diff --git a/android/Android.mk b/android/Android.mk
index 8ce31a4..744c302 100644
--- a/android/Android.mk
+++ b/android/Android.mk
@@ -191,6 +191,11 @@ ifneq ($(strip $(BOARD_MESA3D_GALLIUM_DRIVERS)),)
 # This module also trigger DRI symlinks creation process
 $(eval $(call mesa3d-lib,libgallium_dri,,MESA3D_GALLIUM_BIN))
 
+ifneq ($(filter radeonsi,$(BOARD_MESA3D_GALLIUM_DRIVERS)),)
+# Mesa's GBM-backed legacy DRI frontend used by minigbm's AMD allocator.
+$(eval $(call mesa3d-lib,radeonsi_dri,dri,MESA3D_DRIL_BIN))
+endif
+
 # Module 'libEGL_mesa', produces '/vendor/lib{64}/egl/libEGL_mesa.so'
 $(eval $(call mesa3d-lib,libEGL_mesa,egl,MESA3D_LIBEGL_BIN))
 # Module 'libGLESv1_CM_mesa', produces '/vendor/lib{64}/egl/libGLESv1_CM_mesa.so'
diff --git a/android/mesa3d_cross.mk b/android/mesa3d_cross.mk
index eee4c07..588f8b9 100644
--- a/android/mesa3d_cross.mk
+++ b/android/mesa3d_cross.mk
@@ -70,11 +70,13 @@ $(M_TARGET_PREFIX)MESA3D_LIBEGL_BIN      := $(MESON_OUT_DIR)/install/usr/local/lib/libEGL.so
 $(M_TARGET_PREFIX)MESA3D_LIBGLESV2_BIN   := $(MESON_OUT_DIR)/install/usr/local/lib/libGLESv2.so
 $(M_TARGET_PREFIX)MESA3D_LIBGBM_BIN      := $(MESON_OUT_DIR)/install/usr/local/lib/$(MESA_LIBGBM_NAME).so
 $(M_TARGET_PREFIX)MESA3D_DRI_GBM_BIN     := $(MESON_OUT_DIR)/install/usr/local/lib/gbm/dri_gbm.so
+$(M_TARGET_PREFIX)MESA3D_DRIL_BIN        := $(MESON_OUT_DIR)/install/usr/local/lib/dri/radeonsi_dri.so
 
 
 MESA3D_GBM_BINS := \
     $($(M_TARGET_PREFIX)MESA3D_LIBGBM_BIN) \
     $($(M_TARGET_PREFIX)MESA3D_DRI_GBM_BIN)   \
+    $($(M_TARGET_PREFIX)MESA3D_DRIL_BIN) \
 
 MESA3D_GLES_BINS := \
     $($(M_TARGET_PREFIX)MESA3D_GALLIUM_BIN) \
diff --git a/src/meson.build b/src/meson.build
index 74ea938..69194bd 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -154,7 +154,7 @@ if with_egl
   subdir('egl')
 endif
 if with_gallium
-  if with_glx == 'dri' or with_platform_x11 or with_platform_xcb
+  if with_glx == 'dri' or with_platform_x11 or with_platform_xcb or with_platform_android
     subdir('gallium/targets/dril')
   endif
 endif
